Action permission fixes

mcarcaso opened the pull request:
- Make the auth checks for enabled/available happen in parallel
- Fix a couple of spots where we were checking if the action has isEnabled/isAvailable to not make that  check because it doesn't account for permissions. Best is to just trust the createIsEnabled$ and createIsAvailable$ methods.
- Make createIsAvailable$ and createIsEnabled$ methods context oriented.
- Restructure the Action code a little so there's less of it.

mcarcaso committed:
Remove mention of permissions from ActionView.

mcarcaso committed:
In the absense of an auth server, always allow users to interact with actions.

mcarcaso committed:
Merge branch 'master' into FixExpSlot

mcarcaso committed:
Make Action's createIsEnabled and createIsAvailable context oriented.

mcarcaso committed:
Update calls to createIsEnabled/Available$

mcarcaso commented on src/foam/core/Action.js:
```diff
@@ -134,59 +134,64 @@ foam.CLASS({
-    function checkPermission(x, permission) {
-      if ( ! this.permissionConfig || ! x.auth ) return Promise.resolve(true);
-      return x.auth.check(null, permission);
+    function hasPermissions_(x, data, property) {
```
I don't need data here. Fixing...

mcarcaso committed:
Remove data arg from action permission checks.

kgrgreer commented on src/foam/core/Action.js:
```diff
@@ -134,59 +134,58 @@ foam.CLASS({
+    function createPermissionlessCheckFor_(data$, property) {
+      if ( ! this[property] ) return foam.core.ConstantSlot.create({value: true});
+      return foam.core.ExpressionSlot.create({
+        obj$: data$,
```
I don't think ExpressionSlot supports 'obj' changing, so you could just do obj: data$.get(), which would be more efficient and generate less garbage/listeners.

kgrgreer commented on src/foam/core/Action.js:
```diff
@@ -134,59 +134,58 @@ foam.CLASS({
-
-    function createIsEnabled$(data$) {
-      return this.createSlot_(this.isEnabled, 'enabled', data$);
+    function createPermissionlessCheckFor_(data$, property) {
```
Naming the second argument 'property' is confusing because I expected it would be a Property. Maybe naming something like 'f' or 'fn' would be clearer. And then calling like:
this.createPermissionlessCheckFor_(data$, this.isEnabled)
would be simpler and clearer.

kgrgreer commented on src/foam/core/Action.js:
```diff
@@ -134,59 +134,58 @@ foam.CLASS({
-        await this.hasPermissions_('available', data$)
-      ) {
+      Promise.all([
+        Promise.resolve(this.createPermissionlessCheckFor_(data$, 'isEnabled').get()),
```
Do you need to have Promise.resolve()? Doesn't Promise.all() also accept non-promises?

adamvy commented on src/foam/core/Action.js:
```diff
@@ -134,59 +134,58 @@ foam.CLASS({
+    function createIsEnabled$(x, data$) {
+      return foam.core.ArraySlot.create({
+        slots: [
+          foam.core.PromiseSlot.create({
```
Can we avoid creating the PromiseSlot in the case where we _don't_ have permissions or an auth service?  If we leave it as a ConstantSlot then we can save even more listeners from being hooked up in what is currently a common case of no permissions on an action.

adamvy commented on src/foam/core/Action.js:
```diff
@@ -134,59 +134,58 @@ foam.CLASS({
-        await this.hasPermissions_('enabled', data$) &&
-        await this.hasPermissions_('available', data$)
-      ) {
+      Promise.all([
```
Let's add a comment here noting that because the action is called asynchronously, if it attempts to open any windows with window.open then it will get rejected by the browser's popup blocker.

As you can only do window.open synchronously from an onclick event handler.

It would be extra nice but not as elegant if we avoided the async path in the case where we don't have permissions or we don't have an auth service to check again.

kgrgreer replied:
I agree.

adamvy committed:
Rework Action permission checks:

- No longer async if no permission check is required
- Computes permissions at runtime and based upon data.cls_.id not sourceCls_
- Less code.

adamvy committed:
Improve comments and missing type info for enabledPermissions

adamvy committed:
Add foam.Array.unique
Remove stale entry from nanos.js
Fix Action.maybeCall

adamvy committed:
Merge branch 'master' into FixExpSlot

adamvy committed:
Fix typo

adamvy committed:
Fix logic errors in maybeCall_.  Don't do permission check if there's no permissions.

adamvy committed:
Update calls to createIsEnabled$/createIsAvaialble$ not to pass data as a slot.

If somebody does pass a slot in, unpack it and output a warning.  We were never dynamic on the data$ in createIsEnabled$, but receiving a slot might make people think that we were.

adamvy committed:
Update logic of addPermissionCheck to preserve expected behaviour in nanopay repository:
Fix bug in usage of Promise.all()

Setting enabled/availablePermissions without setting permissionRequired will still invoke a permission check.

Leaving enabled/availablePermissions null and setting permissionRequired to true will compute default permissions.

Setting permissionRequired true and enabledPermissions to an empty array will _not_ invoke a permission check.

adamvy committed:
Fix other usage of Promise.all

adamvy committed:
Remove permissionRequired from Action and simplify resulting code.

kgrgreer commented on src/foam/core/Action.js:
```diff
@@ -113,80 +112,99 @@ foam.CLASS({
-        : true;
-    },
+      var slot = expression ?
+          foam.core.ExpressionSlot.create({
```
I think you can just do data.slot(expression).

adamvy replied:
Done

kgrgreer replied:
It still shows the old code.

kgrgreer commented on src/foam/core/stdlib.js:
```diff
@@ -704,6 +704,17 @@ foam.LIB({
         }
       }
+    },
+    function unique(a, comparator) {
```
Add docs.

adamvy replied:
done

adamvy committed:
Use data.slot instead of ExpressionSlot.create.  Shorter and more readable.

adamvy committed:
Document foam.Array.unique
